app/models/game.rb

```diff
@@ -30,27 +39,34 @@ def populate_game!
     save
   end
 
-  def check?
-    white_king = pieces.find_by(type: 'King', color: 'white')
-    black_pieces = pieces.where(color: 'black')
-
-    black_pieces.each { |black_piece| return true if black_piece.valid_move?(white_king.x_coordinate, white_king.y_coordinate) }
-
-    black_king = pieces.find_by(type: 'King', color: 'black')
-    white_pieces = pieces.where(color: 'white')
-
-    white_pieces.each { |white_piece| return true if white_piece.valid_move?(black_king.x_coordinate, black_king.y_coordinate) }
+  # Keep track of current king.  We don't care what the other king is doing.
+  def check?(color)
+    king = pieces.find_by(type: 'King', color: color)
+    opposing_pieces = pieces.where(color: king.opposite_color)
 
+    opposing_pieces.each { |opposing_piece| return true if opposing_piece.valid_move?(king.x_coordinate, king.y_coordinate) }
     false
   end
 
-  def inactive_player
-    return player_1_id if turn == player_2_id
-    return player_2_id if turn == player_1_id
+  def checkmate?(color)
+    #  a. king has to be in check
+    return false unless check?(color)
+
+    #  a1. king cannot get out of check (has no more valid moves to get out of check)
+    #  a2. king cannot be blocked by another piece (any color)
+    #  a3. pieces checking king check cannot be captured
+    pieces.find_by(color: color, captured: false).each do |piece|
+      piece.valid_move?(x, y).each do |move|
+        return false unless check_if_move(piece, move.x, move.y)
+      end
+    end
+    true
   end
 
-  def not_your_turn
-    !inactive_player
+  # opposing pieces check if king is in_check_if it moves towards the king
+  def check_if_move?(moving_piece, move_to_x, move_to_y)
+    moving_piece.move_to(move_to_x, move_to_y)
```

Comment:
We had discussed using non-mutated move_to instead of using Active Records .Transactions, which seemed to be the right way to go.  This code still doesn't sit right with me.  Busy week for me.  Will revisit it later.